go/apps/mixnet/conn.go

```diff
@@ -42,17 +47,49 @@ func (c Conn) Read(msg []byte) (n int, err error) {
 	return n, nil
 }
 
-// Write a cell to the channel. If the len(msg) > CellBytes, return an error.
-// If len(msg) < CellBytes, then zero-pad the message up to length of a cell.
-func (c Conn) Write(msg []byte) (n int, err error) {
-	cell := make([]byte, CellBytes) // Initialzed as an array of 0-bytes.
-	if len(msg) > CellBytes {
+// Write a cell to the channel. If the len(cell) != CellBytes, return an error.
+func (c *Conn) Write(msg []byte) (n int, err error) {
+	if len(msg) != CellBytes {
 		return 0, errCellLength
 	}
-	copy(cell, msg)
-	n, err = c.Conn.Write(cell)
+	n, err = c.Conn.Write(msg)
 	if err != nil {
 		return n, err
 	}
 	return n, nil
 }
+
+// padCell pads a message to the length of a cell and returns a byte slice of
+// length CellBytes. If len(msg) > CellBytes, then return msg.
+func padCell(msg []byte) []byte {
+	if len(msg) >= CellBytes {
+		return msg
+	}
+
+	cell := make([]byte, CellBytes) // Initially zero
+	bytes := copy(cell, msg)
+	cell[bytes] = 0xf0
+	return cell
+}
+
+// unpadCell removes the padding from a message whose length is less than
+// CellBytesa.
+func unpadCell(cell []byte) []byte {
```

Comment:
Just to record what we discussed offline: this can be simplified and fixed to remove the padding/unpadding and keep track of the number of bytes read from cells. Then the receiver knows if it's read all the expected bytes (since the send puts a length at the beginning of the first message).